File: src/api/binance.py
import requests

def get_price_from_binance(symbol: str) -> float | None:
    pair = symbol.upper() + "USDT"
    url = f"https://api.binance.com/api/v3/ticker/price?symbol={pair}"
    try:
        response = requests.get(url, timeout=10)
        response.raise_for_status()
        data = response.json()
        return float(data["price"])
    except requests.exceptions.RequestException as e:
        print(f"[Binance] 查詢 {symbol} 價格失敗：{e}")
        return None

if __name__ == "__main__":
    from rich import print
    print(get_price_from_binance("eth"))

File: src/api/binance.rs
use reqwest::Client;
use serde::Deserialize;

#[derive(Deserialize, Debug)]
struct BinancePrice {
    price: String,
}

pub async fn _get_price_from_binance(symbol: &str) -> Result<f64, String> {
    let pair = format!("{}USDT", symbol.to_uppercase());
    let url = format!("https://api.binance.com/api/v3/ticker/price?symbol={}", pair);

    let client = Client::builder()
        .timeout(std::time::Duration::from_secs(5))
        .build()
        .map_err(|e| e.to_string())?;

    let response = client.get(&url).send().await.map_err(|e| {
        format!("[Binance] Failed to query price for {}: {}", symbol, e)
    })?;

    if response.status().is_success() {
        let data: BinancePrice = response.json().await.map_err(|e| {
            format!("[Binance] Returned JSON format error: {}", e)
        })?;
        data.price.parse::<f64>().map_err(|_| "Failed to parse price as float".to_string())
    } else {
        Err(format!("[Binance] HTTP error code: {}", response.status()))
    }
}
